components/script/dom/event.rs

```diff
@@ -103,6 +103,36 @@ impl Event {
         self.cancelable.set(cancelable);
     }
 
+    // Determine if there are any listeners for a given target and type.
+    // See https://github.com/whatwg/dom/issues/453
+    pub fn has_listeners_for(&self, target: &EventTarget, type_: &Atom) -> bool {
+        // TODO: take 'removed' into account? Not implemented in Servo yet.
+        // https://dom.spec.whatwg.org/#event-listener-removed
+        let mut event_path = self.construct_event_path(&target);
+        event_path.push(DomRoot::from_ref(target));
+        event_path.iter().filter(|target| target.has_listeners_for(type_)).last().is_some()
+    }
+
+    // https://dom.spec.whatwg.org/#event-path
+    fn construct_event_path(&self, target: &EventTarget) -> Vec<DomRoot<EventTarget>> {
+        let mut event_path = vec![];
```

Comment:
In the old code we were doing a `rooted_vec!` to store the elements in the event path, I'm not sure if we can use a `Vec` here.

cc @jdm

Reply:
I'm not really clear on why we were using a rooted vector previously. From a safety perspective this change is fine because we're now storing DomRoot values in the vector instead of unrooted Dom values.

Reply:
Ok. By the way we are still using a `rooted_vec!` there: 

https://github.com/servo/servo/pull/20329/files#diff-b9c83174d4ef992d8bc1914c8826e220R165

While I had this new `construct_event_path` return a `Vec<DomRoot<EventTarget>>` so that the caller would still have the option to 'transform' it into a rooted vev if necessary, while this seemed unncessary inside `has_listeners_for`, see https://github.com/servo/servo/pull/20329/files#diff-b9c83174d4ef992d8bc1914c8826e220R111.

Perhaps we can also remove [that call to `rooted_vec!`](https://github.com/servo/servo/pull/20329/files#diff-b9c83174d4ef992d8bc1914c8826e220R165) and simply pass the `Vec<DomRoot<EventTarget>>` to [`dispatch_to_listeners`](https://github.com/servo/servo/pull/20329/files#diff-b9c83174d4ef992d8bc1914c8826e220R167)